pkg/util/validation/stack.go

```diff
@@ -20,16 +20,33 @@ import (
 	"regexp"
 
 	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/tokens"
 )
 
-// validateStackName checks if s is a valid stack name, otherwise returns a descriptive error.
+// ValidateStackName checks if s is a valid stack name, otherwise returns a descriptive error.
 // This should match the stack naming rules enforced by the Pulumi Service.
-func validateStackName(s string) error {
-	stackNameRE := regexp.MustCompile("^[a-zA-Z0-9-_.]{1,100}$")
-	if stackNameRE.MatchString(s) {
-		return nil
+func ValidateStackName(s string) error {
+	if s == "" {
+		return errors.New("a stack name may not be empty")
+	}
+	if len(s) > 100 {
+		return errors.New("a stack name cannot exceed 100 characters")
+	}
+
+	stackNameRE := regexp.MustCompile("^" + tokens.NameRegexpPattern)
```

Comment:
Is this missing $ at the end?

Reply:
Ah no I see the len check

Reply:
Seeing as this regex is fixed, instead of compiling it on each ValidateStackName call, we should move it into a top-level private var.

```
var stackNameRegex = regex.MustCompile(..)

func ValidateStackName(..) error {
  // ...
}
```